oauth2_http/javatests/com/google/auth/oauth2/OAuth2CredentialsTest.java

```diff
@@ -348,6 +397,211 @@ public void getRequestMetadata_temporaryToken_hasToken() throws IOException {
+    TestUtils.assertContainsBearerToken(
+        creds.getRequestMetadata(CALL_URI), refreshedToken.getTokenValue());
+  }
+
+  @Test
+  public void getRequestMetadata_staleTemporaryToken_expirationWaits() throws Throwable {
+    Calendar calendar = Calendar.getInstance();
+    Date actualExpiration = calendar.getTime();
+
+    calendar.setTime(actualExpiration);
+    calendar.add(
+        Calendar.MILLISECOND, -1 * Ints.checkedCast(OAuth2Credentials.REFRESH_MARGIN_MILLISECONDS));
+    Date clientStale = calendar.getTime();
+
+    calendar.setTime(actualExpiration);
+    calendar.add(
+        Calendar.MILLISECOND, -1 * Ints.checkedCast(OAuth2Credentials.MINIMUM_TOKEN_MILLISECONDS));
+    Date clientExpired = calendar.getTime();
+
+    TestClock testClock = new TestClock();
+
+    // Initialize credentials which are initially stale and set to refresh
+    final SettableFuture<AccessToken> refreshedTokenFuture = SettableFuture.create();
+    OAuth2Credentials creds =
+        new OAuth2Credentials(new AccessToken(ACCESS_TOKEN, actualExpiration)) {
+          @Override
+          public AccessToken refreshAccessToken() {
+
+            try {
+              return refreshedTokenFuture.get();
+            } catch (Exception e) {
+              throw new RuntimeException(e);
+            }
+          }
+        };
+    creds.clock = testClock;
+
+    // Calls should return immediately with stale token, but a refresh is scheduled
+    testClock.setCurrentTime(clientStale.getTime());
+    MockRequestMetadataCallback callback = new MockRequestMetadataCallback();
+    creds.getRequestMetadata(CALL_URI, realExecutor, callback);
+    TestUtils.assertContainsBearerToken(callback.metadata, ACCESS_TOKEN);
+
+    // Fast forward to expiration, which will hang cause the callback to hang
+    testClock.setCurrentTime(clientExpired.getTime());
+    // Make sure that the callback is hung (while giving it a chance to run)
+    for (int i = 0; i < 10; i++) {
+      Thread.sleep(10);
+      callback = new MockRequestMetadataCallback();
+      creds.getRequestMetadata(CALL_URI, realExecutor, callback);
+      assertNull(callback.metadata);
+    }
+
+    // Resolve the outstanding refresh
+    AccessToken refreshedToken =
+        new AccessToken(
+            "2/MkSJoj1xsli0AccessToken_NKPY2",
+            new Date(testClock.currentTimeMillis() + TimeUnit.HOURS.toMillis(1)));
+    refreshedTokenFuture.set(refreshedToken);
+
+    // The access token should available once the refresh thread completes
+    TestUtils.assertContainsBearerToken(
+        creds.getRequestMetadata(CALL_URI), refreshedToken.getTokenValue());
+    callback = new MockRequestMetadataCallback();
+    creds.getRequestMetadata(CALL_URI, realExecutor, callback);
+    TestUtils.assertContainsBearerToken(callback.awaitResult(), refreshedToken.getTokenValue());
+  }
+
+  @Test
+  public void getRequestMetadata_singleFlightErrorSharing()
```

Comment:
what is the meaning of the "singleFlightErrorSharing"? The test makes sense, but unfamiliar with the terminology

Reply:
Internal link is go/singleflight, but it seems to exist in public domain as well. Single flight is a mechanic to coalesce multiple requests into one. A general implementation would allow for batching, but in this case we only have a single operation type that we need to combine

Reply:
Thanks for pointers :)